[Prepare for go1.20] *: Bump versions and fix tests

MadhavJivrajani opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Commit 1:

Bumping version of govmomi to include changes that
better handle TLS errors. Bump nescessary
to prepare for when the version of Go is
bumped to 1.20.

Command run:
```
./hack/pin-dependency.sh github.com/vmware/govmomi v0.30.0 
```

See https://github.com/kubernetes/kubernetes/pull/114502#issuecomment-1369026247

Commit 2:

This PR also adapts to this bump.

Commit 3:

This commit modifies the unwrap logic of tests for vclib connection tests to accommodate for new error types introduced in go1.20 and be backward compatible with go1.19

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

#### Special notes for your reviewer:

~On running the canary job for unit tests, the `TestWithVerificationWithoutCaCertOrThumbprint` test will still fail. See https://github.com/kubernetes/kubernetes/pull/114502#issuecomment-1369030189~

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/priority important-soon
/assign @liggitt @dims 

(I messed up something in git because of which the previous PR closed, apologies!)

MadhavJivrajani committed:
*: Bump version of vmware/govmomi

Bumping version to include changes that
better handle TLS errors. Bump nescessary
to prepare for when the version of Go is
bumped to 1.20

Signed-off-by: Madhav Jivrajani <madhav.jiv@gmail.com>

MadhavJivrajani committed:
vsphere: Adapt to govmomi version bumps

Signed-off-by: Madhav Jivrajani <madhav.jiv@gmail.com>

MadhavJivrajani commented:
/test pull-kubernetes-unit-go-canary

The `TestWithVerificationWithoutCaCertOrThumbprint` test will still fail.
See https://github.com/kubernetes/kubernetes/pull/114502#issuecomment-1369030189

liggitt commented:
xref https://github.com/kubernetes/release/issues/2815

liggitt commented:
hmm... this thinks total dependencies increased by 4:

```diff
 {
-	"directDependencies": 212,
-	"transitiveDependencies": 234,
-	"totalDependencies": 285,
+	"directDependencies": 213,
+	"transitiveDependencies": 238,
+	"totalDependencies": 289,
 	"maxDepthOfDependencies": 23
 }
```

liggitt commented:
/test pull-kubernetes-unit-go-canary

liggitt commented:
hmm... govmomi already includes 6700+ init() functions (80+% of the init functions in kubernetes), and this library bump adds ~470+ more (search "+func init()" in https://patch-diff.githubusercontent.com/raw/kubernetes/kubernetes/pull/114766.diff)

the unfortunate thing is that those run on every startup of every binary they are linked into, even ones which don't make use of vmware functionality at all :-/

liggitt commented:
new deps this adds to the dependency tree:
* github.com/a8m/tree
* github.com/dougm/pretty
* github.com/pkg/diff
* github.com/vmware/vmw-guestinfo

deps this drops from the dependency tree:
* gopkg.in/errgo.v2

fedebongio commented:
/triage accepted

liggitt commented:
/test check-dependency-stats

liggitt commented:
After chasing https://github.com/kubernetes/kubernetes/pull/114766#issuecomment-1370164010, I think that's spurious... need to dig more, but I don't think that blocks this merging

I'm not thrilled with https://github.com/kubernetes/kubernetes/pull/114766#issuecomment-1370168605, especially since we'll likely need to backport this to release branches in order to move them to go1.20. That said, I don't see a quick fix to that.

MadhavJivrajani commented:
/test pull-kubernetes-unit-go-canary

MadhavJivrajani committed:
vclib: Modify x509.UnknownAuthorityError unwrap check

Modify unwrap error utility to make it work with go1.20
This version of Go introduces a new layer of wrapping via
a new error type. The commit accounts for that while being
compatible with go1.19

Signed-off-by: Madhav Jivrajani <madhav.jiv@gmail.com>

MadhavJivrajani commented:
/test pull-kubernetes-unit-go-canary

MadhavJivrajani commented:
/test pull-kubernetes-unit-go-canary

MadhavJivrajani commented:
@liggitt done.
/test pull-kubernetes-unit-go-canary

liggitt commented:
both unit tests flaked on https://github.com/kubernetes/kubernetes/pull/114678#issuecomment-1374075860

/retest

MadhavJivrajani commented:
/retest

liggitt commented:
/test pull-kubernetes-dependencies
to run check added in https://github.com/kubernetes/kubernetes/pull/114952

liggitt commented:
/lgtm
/approve

MadhavJivrajani commented on staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_test.go:
```diff
@@ -165,7 +165,8 @@ func configFromSimWithTLS(tlsConfig *tls.Config, insecureAllowed bool) (VSphereC
 
 	// vAPI simulator
-	path, handler = vapi.New(s.URL, vpx.Setting)
+	paths, handler := vapi.New(s.URL, vpx.Setting)
+	path = paths[0]
```
I had discarded them, iirc according to my understanding `paths[0]` is what was initially returned from `vapi.New` before the API was changed:

https://github.com/vmware/govmomi/blob/2bd857b669cb777eb13aaa90e9d49b3f2e39081a/vapi/simulator/simulator.go#L182

The commit that made this change: https://github.com/vmware/govmomi/commit/2017e846082636c8910a54f9ffaec5c31475e8b4

liggitt commented:
/lgtm
